cypress-io/cypress
# chore: upgrade lerna to 6, cache build step

- Closes #26798

### Additional details
This PR does the following:

- Upgrades Lerna from `5` => `6`
- Adds `build` as cacheable step

It allows for the `build` steps for the following packages to be cached locally and in the distributed cache (using [Nx Cloud](https://cloud.nx.app/orgs/644c098e2fff8b000dd18049/workspaces/645a6ebd0dd02c0868648ea3/runs)):
- `cypress` (`cli/`)
- `@cypress/angular`
- `create-cypress-tests`
- `@cypress/schematic`
- `@cypress/mount-utils`
- `@cypress/react`
- `@cypress/react18`
- `@cypress/svelte`
- `@cypress/vite-dev-server`
- `@cypress/vite-plugin-cypress-esm`
- `@cypress/vue`
- `@cypress/vue2`
- `@cypress/webpack-dev-server`
- `@cypress/webpack-preprocessor`
- `@packages/electron`
- `@packages/errors`
- `@packages/example`
- `@packages/extension`
- `@packages/icons`
- `@packages/packherd-require`
- `@packages/runner`
- `@packages/telemetry`
- `@packages/v8-snapshot-require`
- `@tooling/electron-mksnapshot`
- `@tooling/packherd`
- `@tooling/v8-snapshot`

#### Rationale

Most of the time when we work on the Cypress repo, we are only editing files in a few packages. If we've already built the source code for the other packages that we **did not** edit, then we shouldn't waste time and compute resources building all of those packages again. 

We can use Lerna (using Nx under the hood) to cache builds when there have been no source code changes in the package. This way, when you run `yarn` or `yarn build`, Lerna only builds the packages that we haven't already built with that exact source code in the past.

As of right now we are being cautious to avoid accidental false positive cache hits, so we have configured Nx to invalidate the build cache for a package when **any** source files in the package change. In the future, and as we cache more steps, we should be able to fine-tune this so that we can exclude files that don't affect build outputs (`*.md` files, `.eslintignore`, etc.).

You can tell if Nx built your package from cache or not based on the output. To the right of the command, you will see some information in brackets like this

```
> @cypress/react18:build  [existing outputs match the cache, left as is]
```

This tells us that Nx did not actually build the `@cypress/react18` package, but it pulled the build output from cache, because we've built `@cypress/react18` with this exact source code before.

To better understand Nx features related to caching, you can read the following
- [Nx task result caching ](https://nx.dev/core-features/cache-task-results)
- [Remote caching with Nx Cloud](https://nx.dev/core-features/remote-cache)
- [Distributed task execution](https://nx.dev/core-features/distribute-task-execution)

#### Escape hatches

If you are building the Cypress monorepo and Nx is incorrectly pulling a build from cache, you can run the command with the `--skip-nx-cache` flag, or set `NX_SKIP_NX_CACHE=true` in your environment to force Nx to build everything from source and to ignore the cache. This normally should not happen, and if it does, then it means that we need to update the Nx configuration in the repo.

#### Going forward

We still have a bunch of packages that are built and tasks that are run using Gulp. These can't be orchestrated or cached by Nx. There are lots of performance gains to be made by moving these tasks out of Gulp and executing them via yarn scripts with Nx. In the future, we will slowly move these tasks out of Gulp and use Nx to orchestrate and cache them.

### Steps to test


Run `yarn` locally on this branch. A lot of the packages that we build during the postinstall step should be taken from the remote cache if you haven't changed any code since CI has already built these packages and uploaded the results to Nx Cloud.

Make a change to the source code of a package from the list above, then run `yarn` again, and verify that the package was **not** built from cache but was actually built with your new source code changes.

### How has the user experience changed?


n/a

### PR Tasks


- [na] Have tests been added/updated?
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
